related to NLO PDF reweighting:

qliphy opened the pull request:
See [1].   Gridpack producing is totally ok. Instead, the problem appears when testing NLO gridpack to produce events.  And it seems to be related to MG242/madgraph/various/banner.py 

The proposed resolution here is to simply remove "re.VERBOSE" in banner.py , and this has been tested with wplustest_5f_NLO successfully

To see the problem more clearly, you can play with the following python code.  Note in MG25X afterwards, banner.py  has been modified a lot.

Would appreciate if someone recheck on this.

--------------------------
#process/bin/internal/banner.py
#def __setitem__(self, name, value, change_userdefine=False):

import re
value= "306000, 322500, 322700, 322900, 323100, 323300, 323500, 323700, 323900, 305800, 13000, 13065, 13069, 13100, 13163, 13167, 13200, 25200, 25300, 25000, 42780, 90200, 91200, 90400, 91400, 61100, 61130, 61200, 61230, 13400, 82200, 292200, 292600, 315000, 315200, 262000, 263000"

value = filter(None, re.split(r'(?:(?<!\\)\s)|,', value, re.VERBOSE))
print "value=", value
------------------------


[1] https://hypernews.cern.ch/HyperNews/CMS/get/prep-ops/4444.html

danbarto commented:
I can confirm that after (manually) patching banner.py running the test command for the gridpack works. I had to patch both `banner.py` files in `mgbasedir/madgraph/various/` and `process/bin/internal/`.

It seems like there's a more fundamental bug in re though, even for the very basic commands:
```
import re
v = '1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36'
re.split(r' |,', v, re.X) #fails
re.split(r' |,', v) #works
```

jfernan2 commented:
It works for me too in ttbar FxFx gridpack. However I get a warning:
WARNING: Failed to access python version of LHAPDF: cannot compute PDF uncertainty from the weights in the events. The weights in the LHE event files will still cover all PDF set members, but there will be no PDF uncertainty printed in the run summary.
 If the python interface to LHAPDF is available on your system, try adding its location to the PYTHONPATH environment variable and theLHAPDF library location to LD_LIBRARY_PATH (linux) or DYLD_LIBRARY_PATH (mac os x).

danbarto commented:
Turns out re.split has no option for re.VERBOSE: re.split(pattern, string, maxsplit=0, flags=0).
re.VERBOSE (same as re.X) has a numerical value of 64, hence the problem.

qliphy commented:
@danbarto  Thanks!  

>"I had to patch both banner.py files in mgbasedir/madgraph/various/ and process/bin/internal/."

If you regenerate gridpack, only the former is needed, as the latter will be a copy.

qliphy commented:
@jfernan2  Thanks. However, I don't see the LHAPDF warning you mentioned. 

 I only got 3 warnings as: "WARNING: Could not access LHAPDF to compute uncertainties for HERAPDF20_NLO_VAR ..."
But this was mentioned in https://github.com/cms-sw/genproductions/pull/1243 and seems to be harmless

qliphy commented:
As can be seen in https://docs.python.org/3/library/re.html  and as also mentioned by @danbarto ,   re.split is defined as re.split(pattern, string, maxsplit=0, flags=0),  so indeed one needs to add "flags=" in banner.py, i.e. 

change from
value = filter(None, re.split(r'(?:(?<!\)\s)|,', value, re.VERBOSE))
to
value = filter(None, re.split(r'(?:(?<!\)\s)|,', value, flags=re.VERBOSE))

I have tested with wplustest_5f_NLO and it works

@kdlong @ALL Can you have a look?  Shall we go with this?   It will affect the NLO FxFx gridpacks

kdlong commented:
Thanks Qiang. Can you push this change and then I'll test? Probably also a good idea to add an FxFx example card.

qliphy commented:
@kdlong   The PR has been updated.   For the cards, cards/examples/wplustest_5f_NLO should be fine.

kdlong commented:
I thought you said it needed to be an FxFx sample?

kdlong commented:
But I thought you said the issue only affected FxFx samples?

danbarto commented:
@kdlong The issue should have affected all NLO samples since it was related to the length of the PDF list, e.g. I had problems with samples without FxFx.

qliphy commented:
@kdlong  My mistake,  not only for FxFx but in general for NLO.  Thanks @danbarto

kdlong commented:
Tested and working for me, I suggest we merge. Thanks all!
